presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.project;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static com.facebook.presto.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is intended to be used right after the PageProcessor to ensure
+ * that the size of the pages returned by FilterAndProject and ScanFilterAndProject
+ * is big enough to do not introduce considerable synchronization overhead
+ * <p>
+ * As long as the input page contains more than {@link MergingPageOutput#DEFAULT_MIN_ROW_COUNT} rows
+ * or bigger than {@link MergingPageOutput#DEFAULT_MIN_PAGE_SIZE_IN_BYTES} it is returned as is without
+ * additional memory copy.
+ * <p>
+ * The page data that has been buffered so far before receiving a "big" page is being flushed
+ * before transferring a "big" page.
+ * <p>
+ * Although it is still possible that the {@link MergingPageOutput} may return a tiny page,
+ * this situation is considered to be rare due to the assumption that filter selectivity may not
+ * vary a lot based on the particular input page.
+ * <p>
+ * Considering the CPU time required to process(filter, project) a full (~1MB) page returned by a
+ * connector, the CPU cost of memory copying (< 50kb, < 1024 rows) is supposed to be negligible
+ */
+public class MergingPageOutput
```

Comment:
[TODO] I would prefer `MergingPageOutput` to be streamlined with `PageProcessor` and `PageSource`, e.g: similarly as `Iterator` can be streamlined. I have created an issue for that: https://github.com/prestodb/presto/issues/8697. In such case  `MergingPageOutput`, `PageProcessor` and `PageSource` would implement `BlockingSource<Page>` and they would be chained together, e.g:
`new MergingPageOutput(..., new PageProcessor(..., new PageSource(...)))`

Otherwise, `MergingPageOutput` could be just a base `Operator` class from which other operators (that wan't to merge pages) derive. This way you wouldn't have to reimplement similar mechanism for https://github.com/prestodb/presto/pull/8663